### PR TITLE
Refactor login flow

### PR DESCRIPTION
**Pull request by richvdh**
Make sure that we have the canonical user_id *before* calling
get_login_tuple_for_user_id.

Replace login_with_password with a method which just validates the password,
and have the caller call get_login_tuple_for_user_id. This brings the password
flow into line with the other flows, and will give us a place to register the
device_id if necessary.

### CODE REVIEW

**Comment by richvdh** [comment]
To be clear: this is the first PR of a series which will introduce device_id tracking. This change in itself shouldn't have any functional effects.

File: synapse/handlers/auth.py [hunk 1]
```diff
@@ -240,11 +239,7 @@ def _check_password_auth(self, authdict, _):
         if not user_id.startswith('@'):
             user_id = UserID.create(user_id, self.hs.hostname).to_string()
 
-        if not (yield self._check_password(user_id, password)):
-            logger.warn("Failed password login for user %s", user_id)
-            raise LoginError(403, "", errcode=Codes.FORBIDDEN)
-
-        defer.returnValue(user_id)
+        return self._check_password(user_id, password)
```
Comment (erikjohnston) [inline synapse/handlers/auth.py]: Missing `yield`, non?

File: synapse/handlers/auth.py [hunk 2]
```diff
@@ -240,11 +239,7 @@ def _check_password_auth(self, authdict, _):
         if not user_id.startswith('@'):
             user_id = UserID.create(user_id, self.hs.hostname).to_string()
 
-        if not (yield self._check_password(user_id, password)):
-            logger.warn("Failed password login for user %s", user_id)
-            raise LoginError(403, "", errcode=Codes.FORBIDDEN)
```
Comment (erikjohnston) [inline synapse/handlers/auth.py]: We're changing the API from `raise` on failed login to returning `bool`? Does this not need a change at the calling site?

File: synapse/handlers/auth.py (hunk 1 above)
Comment (erikjohnston) [inline synapse/handlers/auth.py]: Err, ignore me.

File: synapse/handlers/auth.py (hunk 2 above)
Comment (richvdh) [inline synapse/handlers/auth.py]: No, _check_password is changing from returning `bool` to `raise`, so this code becomes redundant

**Comment by erikjohnston** [comment]
LGTM

(Looks like the `Sytest Dendron (Commit)` is nothing to do with this PR)

**Pull request merged**